File: NeetCode/Python/Arrays/Medium/912.SortAnArray.py
# Given an array of integers nums, sort the array in 
# ascending order and return it.

def sortArray(nums):   
    if len(nums) <= 1:
        return nums

    mid = len(nums)//2

    left, right = nums[:mid], nums[mid:]
    left = sortArray(left)
    right = sortArray(right)

    return merge(left, right)

def merge(left, right):
    result = []
    i = 0
    j = 0

    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    while i < len(left):
        result.append(left[i])
        i += 1

    while j < len(right):
        result.append(right[j])
        j += 1

    return result

print(sortArray([5,2,3,1]))
print(sortArray([5,1,1,2,0,0]))

File: swift/Neetcode/Arrays/Medium/912.SortAnArray.swift
// Given an array of integers nums, sort the array in 
// ascending order and return it.

func sortArray(_ nums: [Int]) -> [Int] {
    if nums.count <= 1 { return nums }

    let mid = nums.count / 2
    let left = sortArray(Array(nums[..<mid]))
    let right = sortArray(Array(nums[mid...]))

    var result = [Int]()
    var l = 0, r = 0

    while l < left.count && r < right.count {
        if right[r] > left[l] {
            result.append(left[l])
            l += 1
        } else {
            result.append(right[r])
            r += 1
        }
    }

    return result + left[l...] + right[r...]
}

print(sortArray([5,2,3,1]))
print(sortArray([5,1,1,2,0,0]))